src/java/org/apache/cassandra/journal/Journal.java

```diff
@@ -0,0 +1,644 @@
+    /**
+     * Looks up a record by the provided id.
+     * <p/>
+     * Looking up an invalidated record may or may not return a record, depending on
+     * compaction progress.
+     * <p/>
+     * In case multiple copies of the record exist in the log (e.g. because of user retries),
+     * the first one found will be returned.
+     *
+     * @param id user-provided record id, expected to roughly correlate with time and go up
+     * @return deserialized record if found, null otherwise
+     */
+    public V read(K id)
+    {
+        EntrySerializer.EntryHolder<K> holder = new EntrySerializer.EntryHolder<>();
+
+        try (ReferencedSegments<K> segments = selectAndReference(id))
+        {
+            for (Segment<K> segment : segments.all())
+            {
+                if (segment.read(id, holder))
+                {
+                    try
+                    {
+                        return valueSerializer.deserialize(holder.key, new DataInputBuffer(holder.value, false), segment.descriptor.userVersion);
+                    }
+                    catch (IOException e)
+                    {
+                        // can only throw if serializer is buggy
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Synchronously write a record to the journal.
+     * <p/>
+     * Blocks until the record has been deemed durable according to the journal flush mode.
+     *
+     * @param id user-provided record id, expected to roughly correlate with time and go up
+     * @param record the record to store
+     * @param hosts hosts expected to invalidate the record
+     */
+    public void write(K id, V record, Set<Integer> hosts)
+    {
+        try (DataOutputBuffer dob = DataOutputBuffer.scratchBuffer.get())
+        {
+            valueSerializer.serialize(record, dob, MessagingService.current_version);
+            ActiveSegment<K>.Allocation alloc = allocate(dob.getLength(), hosts);
+            alloc.write(id, dob.unsafeGetBufferAndFlip(), hosts);
+            flusher.waitForFlush(alloc);
+        }
+        catch (IOException e)
+        {
+            // exception during record serialization into the scratch buffer
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * Asynchronously write a record to the journal. Writes to the journal in the calling thread,
+     * but doesn't wait for flush.
+     * <p/>
+     * Executes the supplied callback on the executor provided,
+     * once the record has been deemed durable according to the journal flush mode.
+     *
+     * @param id user-provided record id, expected to roughly correlate with time and go up
+     * @param record the record to store
+     * @param hosts hosts expected to invalidate the record
+     */
+    public void asyncWrite(K id, V record, Set<Integer> hosts, @Nonnull Executor executor, @Nonnull Runnable onDurable)
```

Comment:
When I see `Runnable` for callback, I tend to ask what happens in the failure case... if I am reading this properly `org.apache.cassandra.journal.Flusher.FlushRunnable#run` sees the failure (so does `org.apache.cassandra.journal.Journal#asyncWrite`), but has no way to notify the blocked logic that the segment couldn't save; this can be a problem as the "chain" of operations going on in Accord never knows that the operation failed and won't ever succeed, so can't properly handle the error case on their side.

We should notify if the section of the segment we are waiting on failed for w/e reason and that durability maybe lost

Reply:
I also know this doesn't exist outside of Accord, but `AsyncChain` helps with this pattern... so could have the following signature

```
public AsyncChain<Void> asyncWrite(K id, V record, Set<Integer> hosts)
```

This allows the caller to control things with regard to the callback better, such as what executor is used (`addCallback(callback, executor)`) or what to do after the option complete.

This does not solve the concern that errors do not propagate to callbacks, that is still an issue, this just changes how callbacks are used

Reply:
This is slightly tricky, with journal, just like commit log, mostly handling those exceptions itself, and in practice mostly choosing to stop the node entirely. I did for now pushed a commit that expects a richer callback, however, with an `onError()` method, currently propagating just the message serialization exceptions to it.

Reply:
Resolving now that we both have the proper callback and propagate the errors.